### PR TITLE
Add support for serializer parameters in plumber block

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #315

### CODE REVIEW

**Comment by schloerke** [comment]
appveyor is having trouble installing jsonlite. not a package issue

**Comment by namelessjon** [comment]
Hi @schloerke 

Thanks for finishing this off.  Sorry, December just turned into a month of busy and distraction, and it was just a little too large a bite to find the time for

File: R/parse-block.R [hunk 1]
```diff
@@ -115,7 +115,7 @@ parseBlock <- function(lineNum, file){
       }
     }
 
-    shortSerMat <- stringi::stri_match(line, regex="^#['\\*]\\s*@(json|html)")
+    shortSerMat <- stringi::stri_match(line, regex="^#['\\*]\\s*@(json|html)([\\s\\(].*)?\\s*$")
```
Comment (schloerke) [inline R/parse-block.R]: change to `(.*)$` at addition of regex

File: R/parse-block.R [hunk 2]
```diff
@@ -126,9 +126,22 @@ parseBlock <- function(lineNum, file){
       if (!is.na(s) && !s %in% names(.globals$serializers)){
         stop("No such @serializer registered: ", s)
       }
+      shortSerAttr <- trimws(shortSerMat[1,3])
+      if (is.na(shortSerAttr)){
```
Comment (schloerke) [inline R/parse-block.R]: might not be necessary with regex change above

File: R/parse-block.R [hunk 3]
```diff
@@ -126,9 +126,22 @@ parseBlock <- function(lineNum, file){
...
+        shortSerAttr <- ""
+      }
+      if(!identical(shortSerAttr, "") && !grepl("^\\(.*\\)$", shortSerAttr, perl=TRUE)){
```
Comment (schloerke) [inline R/parse-block.R]: is `perl=TRUE` required?

File: R/serializer-html.R [hunk 4]
```diff
@@ -1,7 +1,12 @@
 #' @rdname serializers
 #' @export
-serializer_html <- function(){
-  function(val, req, res, errorHandler){
+serializer_html <- function(...) {
```
Comment (schloerke) [inline R/serializer-html.R]: serialize_null should accept arguments. check all other serializers accept arguments and throw if any are provided

File: R/serializer-html.R [hunk 5]
```diff
@@ -1,7 +1,12 @@
...
+  args <- list(...)
+  if (length(args) > 0) {
+    warning("'html' serializer does not interpret extra arguments")
```
[the remark this answers is not in the document]
Comment (schloerke) [reply R/serializer-html.R]: "accept"

File: R/parse-block.R [hunk 6]
```diff
@@ -126,9 +126,22 @@ parseBlock <- function(lineNum, file){
...
+        stopOnLine(lineNum, line, "Supplemental arguments to the serializer must be surrounded by parentheses, as in `#' @json(na='null')`")
+      }
+
+      if (shortSerAttr != ""){
+        # We have an arg to pass in to the serializer
+        argList <- as.list(eval(parse(text=paste0("list", shortSerAttr))))
 
-      # TODO: support arguments to short serializers once they require them.
-      serializer <- .globals$serializers[[s]]()
+        serializer <- do.call(.globals$serializers[[s]], argList)
```
[the remark this answers is not in the document]
Comment (schloerke) [reply R/parse-block.R]: Wrap `do.call` in `tryCatch` for better error reporting.  Remove extra `...` args from functions that don't use them.

```r
tryCatch({
    do.call(ser, argList)
  }, error = function(e) {
    print(line)
    print(lineNum)
    print(e)
  }
)
```

File: R/parse-block.R (hunk 3 above)
Comment (schloerke) [reply R/parse-block.R]: no. it's a very simple regex

File: R/parse-block.R [hunk 7]
```diff
@@ -126,9 +130,23 @@ parseBlock <- function(lineNum, file){
       if (!is.na(s) && !s %in% names(.globals$serializers)){
         stop("No such @serializer registered: ", s)
       }
+      shortSerAttr <- trimws(shortSerMat[1,3])
+      if(!identical(shortSerAttr, "") && !grepl("^\\(.*\\)$", shortSerAttr)){
```
Comment (wch) [inline R/parse-block.R]: It would be good to have tests for edge cases so that we can more safely refactor in the future.

File: tests/testthat/test-parse-block.R [hunk 8]
```diff
@@ -79,4 +79,55 @@ test_that("Block can't contain duplicate tags", {
...
+    expect_warning({
+      parseBlock(length(lines), lines)
+    }, ...)
+  }
+
+  expect_block_fn("#' @serializer html", serializer_html())
+  expect_block_fn("#' @html", serializer_html())
+  expect_block_fn("#' @html()", serializer_html())
+
+  expect_block_warning("#' @serializer html list(key = \"val\")", "does not interpret extra arguments")
```
Comment (wch) [inline tests/testthat/test-parse-block.R]: Maybe use "accept arguments" instead of "interpret extra arguments".

**Comment by apparaokalimireddy** [comment]
Is this fix    in latest docker image trestletech/plumber?
Thanks
Apparao
